proof/crefine/RISCV64/Invoke_C.thy

```diff
@@ -75,24 +75,43 @@ lemma setDomain_ccorres:
        apply (simp add: guard_is_UNIV_def)
       apply simp
       apply wp
-     apply (rule_tac Q="\<lambda>_. all_invs_but_sch_extra and tcb_at' t and sch_act_simple
-                        and (\<lambda>s. rv = ksCurThread s)" in hoare_strengthen_post)
-      apply (wp threadSet_all_invs_but_sch_extra)
-     apply (clarsimp simp:valid_pspace_valid_objs' st_tcb_at_def[symmetric]
-       sch_act_simple_def st_tcb_at'_def o_def weak_sch_act_wf_def split:if_splits)
+     apply (rule_tac Q="\<lambda>_. invs' and tcb_at' t and sch_act_simple
+                        and (\<lambda>s. rv = ksCurThread s)
+                        and (\<lambda>s. \<forall>d p. distinct (ksReadyQueues s (d, p)))"
+                  in hoare_strengthen_post)
+       apply wp
+      apply (fastforce simp: valid_pspace_valid_objs' weak_sch_act_wf_def
+                      split: if_splits)
+     apply (rule_tac Q="\<lambda>_. invs' and tcb_at' t and sch_act_simple
+                            and (\<lambda>s. rv = ksCurThread s \<and> (\<forall>p. t \<notin> set (ksReadyQueues s p)))
+                            and (\<lambda>s. \<forall>d p. distinct (ksReadyQueues s (d, p)))"
+                  in hoare_strengthen_post)
+      apply (wpsimp wp: threadSet_tcbDomain_update_invs')
+     apply fastforce
     apply (simp add: guard_is_UNIV_def)
-   apply (rule_tac Q="\<lambda>_. invs' and tcb_at' t and sch_act_simple
-      and (\<lambda>s. rv = ksCurThread s \<and> (\<forall>p. t \<notin> set (ksReadyQueues s p)))" in hoare_strengthen_post)
-    apply (wp weak_sch_act_wf_lift_linear tcbSchedDequeue_not_queued
-              tcbSchedDequeue_not_in_queue hoare_vcg_imp_lift hoare_vcg_all_lift)
-   apply (clarsimp simp: invs'_def valid_pspace'_def valid_state'_def)
-  apply (fastforce simp: valid_tcb'_def tcb_cte_cases_def
-                         invs'_def valid_state'_def valid_pspace'_def)
+   apply (wp tcbSchedDequeue_not_in_queue  hoare_vcg_all_lift)
+  apply fastforce
   done
 
 lemma active_runnable':
   "active' state \<Longrightarrow> runnable' state"
-  by (fastforce simp: runnable'_def)
+  by fastforce
+
+lemma tcbSchedEnqueue_ksReadyQueues_distinct:
+  "\<lbrace>\<lambda>s. \<forall>d p. distinct (ksReadyQueues s (d, p)) \<and> thread \<notin> set (ksReadyQueues s (d,p))\<rbrace>
+   tcbSchedEnqueue thread
+   \<lbrace>\<lambda>_ s. \<forall>d p. distinct (ksReadyQueues s (d, p))\<rbrace>"
```

Comment:
It's rare, but `ksReadyQueues_distinct` lemmas are sometimes useful without the forall. If it's easy, then I would prefer the following, with a `hoare_vcg_all_lift` where it's called.
```suggestion
  "\<lbrace>\<lambda>s. distinct (ksReadyQueues s (d, p)) \<and> thread \<notin> set (ksReadyQueues s (d,p))\<rbrace>
   tcbSchedEnqueue thread
   \<lbrace>\<lambda>_ s. distinct (ksReadyQueues s (d, p))\<rbrace>"
```

Reply:
Yes, I'd also agree that this form is more useful. The post condition binds `d` and `p` so we're not introducing schematics in the precondition when we're leaving out the quantifier here, and it makes the lemma more often applicable.

Reply:
Done. And yes, just one use of `hoare_vcg_all_lift` needed